### DISCUSSION TITLE
How to make ratelimit selector more suitable for a per domain solutions?

### DISCUSSION THREAD

**Discussion by barianiluca** [rspamd/rspamd#4626] [Q&A] [answered]
Dear,

after understanding that ratelimit module does not follows a top-to-down policy, but all the rates here defined are evaluated in a non-reliable way (https://github.com/rspamd/rspamd/issues/4467#issuecomment-1746488912) there is a clean way to archive something like these?

```
rates {
	1000_smtp_mail_daily_limit_customerdomain_com = {
	  # 1000 mail /24h for user of @customerdomain.com domain
	  selector = 'user.lower.regexp("^[A-Za-z0-9._%+-]+@customerdomain\.com$")';
	  bucket = [
	  {
			burst = 1000;
			rate = "1000 / 24h";
	  }]
	}
	smtp_mail_daily_limit = {
	  # 300 mail /24h for others user authenticated users
	  selector = 'user.lower';
	  bucket = [
	  {
			burst = 300;
			rate = "300 / 24h";
	  }]
	}
[...]
}
```

Basically what we want is that the domain of the rate `1000_smtp_mail_daily_limit_customerdomain_com ` can't be grabbed from a selector of `smtp_mail_daily_limit`.
My first thought was a map of regex, and do something like user.lower.in(map) for the first one and user.lower.not_in(map) for the second one. But i'm not sure if this is the best approach, nor if is possible define a map inside that module.
Any ideas?

Thank you

**Comment by fatalbanana** [comment]
It's possible to use [maps in selectors](https://rspamd.com/doc/configuration/selectors.html#maps-in-transformations) but it's a bit limited; I suppose it's probably necessary to use [custom selectors written in Lua](https://rspamd.com/doc/configuration/selectors.html#own-selectors).

~~~lua
-- /etc/rspamd/rspamd.local.lua

local lua_maps = require 'lua_maps'
local lua_selectors = require "lua_selectors"

local higher_limit_domains = lua_maps.map_add_from_ucl(
  '/etc/rspamd/higher_limit_domains.map',
  'set',
  'Domains whose users get higher sending limits'
)

lua_selectors.register_extractor(rspamd_config, "user_higher", {
  get_value = function(task, args)
    local user = task:get_user()
    if not user then return end
    local domain = user:match('@(.*)')
    if not domain then return end
    if not higher_limit_domains:get_key(domain) then return end
    return user:lower(), 'string'
  end,
  description = 'Usernames with higher sending limits'
})

lua_selectors.register_extractor(rspamd_config, "user_lower", {
  get_value = function(task, args)
    local user = task:get_user()
    if not user then return end
    local domain = user:match('@(.*)')
    if not domain then return end
    if higher_limit_domains:get_key(domain) then return end
    return user:lower(), 'string'
  end,
  description = 'Usernames with default sending limits'
})
~~~

Given this stuff in `/etc/rspamd/rspamd.local.lua` & domains in `/etc/rspamd/higher_limit_domains.map`  you could use `user_higher` and `user_lower` respectively as selectors in multimap configuration.

**Comment by barianiluca** [comment]
First, [fatalbanana](https://github.com/fatalbanana) thank you for the sample code.
I took some weeks to study and try something before to come back. Hopefully with something useful.
I was be able to create the custom selector and the maps. The maps has something like this in it:

```
-- /etc/rspamd/higher_limit_domains.map

mydomain.com
```

the rspamd.local.lua is like the yours above, then in the local.d/ratelimit.conf i tried something like this:

```
[...]
1000_smtp_mail_daily_limit_mydomain_com = {
          #
          # 1000 mail / 24h - Limit for @mydomain.com authenticated user domain
          #
          selector = 'user.lower.in("user_higher")';
          bucket = [
          {
                burst = 1000;
                rate = "1000 / 24h";
          }]
         }
}
[...]
```

But that rate is never catch.
There is something that i wrong? Maybe i can't use the custom selector there?
ty

**Reply by fatalbanana** [reply, marked as answer]
~~~
 selector = 'user_higher';
~~~

**Reply by barianiluca** [reply]
thank you man. very much. with your help i was able to set my sending policy with several rates limit.

**Reply by expertaz-zz** [reply]
@barianiluca could you please share your domain or ip based rates limit conf?
